File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_len , start = 0 , -1
        freq = [-1]*256
        for i in range(0,len(s)) :
            if freq[ord(s[i])] > start :
                start = freq[ord(s[i])]
            freq[ord(s[i])] = i
            max_len = max(max_len , i - start)
        
        return max_len

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        if(n == 0) return 0;
        int max_len = 1;
//         Set<Character> set = new HashSet<>();
//         int left = 0, right = 0;
        
//         while(right < n){
//             while(left < n && set.contains(s.charAt(right))){
//                 set.remove(s.charAt(left));
//                 left++;
//             }
//             set.add(s.charAt(right));
//             max_len = Math.max(max_len,right - left + 1);
//             right++;
//         }        
//         return max_len;
        
        Map<Character,Integer> map = new HashMap<>();
        int left = 0 , right = 0;
        
        while(right < n){
            if(map.get(s.charAt(right)) != null)
                left = Math.max((map.get(s.charAt(right)) + 1) , left);

            map.put(s.charAt(right),right);
            max_len = Math.max(max_len,right - left + 1);
            right++;
        }
        return max_len;
    }
}

